vm/systemSmartContracts/governance.go

```diff
@@ -517,236 +748,468 @@ func (g *governanceContract) hardForkProposal(args *vmcommon.ContractCallInput)
+
+	generalProposal := &GeneralProposal{}
+	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
 	if err != nil {
-		log.Warn("saveGeneralProposal", "err", err)
-		g.eei.AddReturnMessage("saveGeneralProposal" + err.Error())
-		return vmcommon.UserError
+		return nil, err
 	}
 
-	return vmcommon.Ok
+	return generalProposal, nil
 }
 
-func (g *governanceContract) vote(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	if args.CallValue.Cmp(zero) != 0 {
-		g.eei.AddReturnMessage("invalid proposal cost, expected 0")
-		return vmcommon.OutOfFunds
-	}
-	err := g.eei.UseGas(g.gasCost.MetaChainSystemSCsCost.Vote)
+// proposalExists returns true if a proposal already exists
+func (g *governanceContract) proposalExists(reference []byte) bool {
+	key := append([]byte(proposalPrefix), reference...)
+	marshaledData := g.eei.GetStorage(key)
+	return len(marshaledData) > 0
+}
+
+// getValidProposal returns a proposal from storage if it exists or it is still valid/in-progress
+func (g *governanceContract) getValidProposal(reference []byte) (*GeneralProposal, error) {
+	proposal, err := g.getGeneralProposal(reference)
 	if err != nil {
-		g.eei.AddReturnMessage("not enough gas")
-		return vmcommon.OutOfGas
+		return nil, err
 	}
-	if len(args.Arguments) < 2 || len(args.Arguments) > 3 {
-		g.eei.AddReturnMessage("invalid number of argument expected 2 or 3")
-		return vmcommon.FunctionWrongSignature
+
+	currentNonce := g.eei.BlockChainHook().CurrentNonce()
+	if currentNonce < proposal.StartVoteNonce {
+		return nil, vm.ErrVotingNotStartedForProposal
 	}
-	if len(args.Arguments) == 3 && len(args.Arguments[2]) != len(args.CallerAddr) {
-		g.eei.AddReturnMessage("wrong argument number 3 should be a valid address")
-		return vmcommon.FunctionWrongSignature
+
+	if currentNonce > proposal.EndVoteNonce {
+		return nil, vm.ErrVotedForAnExpiredProposal
 	}
-	if len(args.Arguments) == 3 && bytes.Equal(args.CallerAddr, args.Arguments[2]) {
-		g.eei.AddReturnMessage("wrong argument number 3 should be different than caller")
-		return vmcommon.FunctionWrongSignature
+
+	return proposal, nil
+}
+
+// isWhiteListed checks if an address is whitelisted
+func (g *governanceContract) isWhiteListed(address []byte) bool {
+	key := append([]byte(whiteListPrefix), address...)
+	marshaledData := g.eei.GetStorage(key)
+	if len(marshaledData) == 0 {
+		return false
 	}
 
-	proposalToVote := args.Arguments[0]
-	if !g.proposalExists(proposalToVote) {
-		g.eei.AddReturnMessage("proposal does not exists")
-		return vmcommon.UserError
+	key = append([]byte(proposalPrefix), address...)
+	marshaledData = g.eei.GetStorage(key)
+	generalProposal := &GeneralProposal{}
+	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
+	if err != nil {
+		return false
 	}
 
-	voteString := string(args.Arguments[1])
-	if !g.isValidVoteString(voteString) {
-		g.eei.AddReturnMessage("argument 1 is not a valid vote string")
+	return generalProposal.Voted
```

Comment:
is this ok? Shouldn't the function return ok if the whiteListPrefix+address contains some marshal data there?

Reply:
no. it is ok.

Reply:
ok